File: InsertionAndBubbleSorts.cpp
#include <iostream>
#include <vector>
#include <chrono>

long long last = 42;
const long long a = 25214903917LL;
const long long c = 11;
const long long m = 1LL << 48;

int nextInt(int bound) {
    last = (a * last + c) % m;
    return last % bound;
}

int partition(std::vector<int>& arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;
    for (int j = low; j < high; ++j) {
        if (arr[j] < pivot) {
            ++i;
            std::swap(arr[i], arr[j]);
        }
    }
    std::swap(arr[i + 1], arr[high]);
    return i + 1;
}

void bubbleSort(std::vector<int>& arr, int n)
{
    int i, j;
    bool swapped;
    for (i = 0; i < n - 1; i++) {
        swapped = false;
        for (j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                std::swap(arr[j], arr[j + 1]);
                swapped = true;
            }
        }

        // If no two elements were swapped
        // by inner loop, then break
        if (swapped == false)
            break;
    }
}

void insertionSort(std::vector<int>& arr, int n)
{
    int i, key, j;
    for (i = 1; i < n; i++) {
        key = arr[i];
        j = i - 1;

        // Move elements of arr[0..i-1],
        // that are greater than key,
        // to one position ahead of their
        // current position
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
    }
}

int main() {
    const int MAX_N = 6000;

    long long initialLast = last;

    for(int N = 100; N <= MAX_N; N *= 1.5) {

        std::vector<int> arr(N);

        const int M = 10;
        std::vector<long long> sortingTimes(M);

        for (int i = 1; i <= M; ++i) {
            last = i;

            for (int j = 0; j < N; ++j) {
                arr[j] = nextInt(1007);
            }

            auto sortStart = std::chrono::high_resolution_clock::now();

            bubbleSort(arr, N);

            auto sortStop = std::chrono::high_resolution_clock::now();
            long long sortDuration = std::chrono::duration_cast<std::chrono::nanoseconds>(sortStop - sortStart).count();

            sortingTimes[i - 1] = sortDuration;
        }

        long long averageTime = 0;

        for (int i = 0; i < M; ++i) {
            averageTime += sortingTimes[i];
        }
        long averageTimeInNanoseconds = averageTime / M;
        // std::cout << "Average Time for array << averageTimeInNanoseconds << " nanoseconds" << std::endl;
        std::cout << averageTimeInNanoseconds / 1000 << std::endl;
    }
    return 0;
}

File: InsertionAndBubbleSorts.py
import time

last = 42
a = 25214903917
c = 11
m = 1 << 48


def next_int(bound):
    global last
    last = (a * last + c) % m
    return last % bound


def bubble_sort(arr):
    n = len(arr)
    for i in range(n - 1):
        for j in range(0, n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]


# Function to do insertion sort
def insertionSort(arr):
    # Traverse through 1 to len(arr)
    for i in range(1, len(arr)):

        key = arr[i]

        # Move elements of arr[0..i-1], that are
        # greater than key, to one position ahead
        # of their current position
        j = i - 1
        while j >= 0 and key < arr[j]:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key


if __name__ == "__main__":
    MAX_N = 6000
    N = 100

    M = 10
    sorting_times = [0] * M

    start = time.time()

    initial_last = last

    while N < MAX_N:
        arr = [0] * N
        for i in range(1, M + 1):
            last = i

            for j in range(N):
                arr[j] = next_int(1007)

            sort_start = time.time()
            bubble_sort(arr)
            sort_stop = time.time()

            sort_duration = (sort_stop - sort_start) * 1e9

            sorting_times[i - 1] = sort_duration

        average_time = sum(sorting_times) / M
        average_time //= 1000
        print(average_time)
        N *= 1.5
        N = int(N)
